Critical bugfixes for AdaptiveIndexDefrag

LetsDoSQL-jdw opened the pull request:
v1.6.8 - 2024-08-30
* Bugfix update statistics because of quotenames, as programmed in https://github.com/AndrewG2/tigertoolbox/commit/00fef9a645e7f6880557cf59692b59f6da1a58c1

v1.6.9 - 2024-08-30
* Adding schema-names in the debug output & logging output

v1.7.0 - 2024-09-02
* Fix update-stats for tables with more than 10B changes since last update-stats
* Fix duration calculation
* Fix logging update-stats of table id, schema name and table name

v1.7.1 - 2024-09-10
* Fix for bug in official release : proactive fix for rebuilding indexes which have spaces in index names
* Fix for bug introduced in v1.7.0: for those indexes rebuilt, the stats_log table had a null-value for objectId & objectName

LetsDoSQL-jdw committed:
Update usp_AdaptiveIndexDefrag to v1.7.1

Most importantly:
- Inspired by bugfix update statistics because of quotenames, as programmed in https://github.com/AndrewG2/tigertoolbox/commit/00fef9a645e7f6880557cf59692b59f6da1a58c1
- Fix long duration calculation
- Fix statistics modification counter for tables exceeding 10B changes since last update-stats
- Remove unnecessary brackets in the logging
- Adding schema names in the logging output & debug output

LetsDoSQL-jdw committed:
Update CHANGELOG.txt

LetsDoSQL-jdw commented:
@microsoft-github-policy-service agree company="LetsDoSQL"

On 2024-09-11 11:53, microsoft-github-policy-service[bot] wrote:

> @LetsDoSQL-jdw [1] please read the following Contributor License 
> Agreement(CLA). If you agree with the CLA, please reply with the 
> following information.
> 
> @microsoft-github-policy-service agree [company="{your company}"]
> 
>> Options:
>> 
>> * (default - no company specified) I have sole ownership of 
>> intellectual property rights to my Submissions and I am not making 
>> Submissions in the course of work for my employer.
>> 
>> @microsoft-github-policy-service agree
>> 
>> * (when company given) I am making Submissions in the course of work 
>> for my employer (or my employer has intellectual property rights in my 
>> Submissions by contract or applicable law). I have permission from my 
>> employer to make Submissions and enter into this Agreement on behalf 
>> of my employer. By signing below, the defined term "You" includes me 
>> and my employer.
>> 
>> @microsoft-github-policy-service agree company="Microsoft"
> Contributor License Agreement
> 
> CONTRIBUTION LICENSE AGREEMENT
> 
> This Contribution License Agreement ("Agreement") is agreed to by the 
> party signing below ("You"),
> and conveys certain license rights to Microsoft Corporation and its 
> affiliates ("Microsoft") for Your
> contributions to Microsoft open source projects. This Agreement is 
> effective as of the latest signature
> date below.
> 
> * Definitions.
> "Code" means the computer software code, whether in human-readable or 
> machine-executable form,
> that is delivered by You to Microsoft under this Agreement.
> "Project" means any of the projects owned or managed by Microsoft and 
> offered under a license
> approved by the Open Source Initiative (www.opensource.org [2]).
> "Submit" is the act of uploading, submitting, transmitting, or 
> distributing code or other content to any
> Project, including but not limited to communication on electronic 
> mailing lists, source code control
> systems, and issue tracking systems that are managed by, or on behalf 
> of, the Project for the purpose of
> discussing and improving that Project, but excluding communication that 
> is conspicuously marked or
> otherwise designated in writing by You as "Not a Submission."
> "Submission" means the Code and any other copyrightable material 
> Submitted by You, including any
> associated comments and documentation.
> * Your Submission. You must agree to the terms of this Agreement before 
> making a Submission to any
> Project. This Agreement covers any and all Submissions that You, now or 
> in the future (except as
> described in Section 4 below), Submit to any Project.
> * Originality of Work. You represent that each of Your Submissions is 
> entirely Your original work.
> Should You wish to Submit materials that are not Your original work, 
> You may Submit them separately
> to the Project if You (a) retain all copyright and license information 
> that was in the materials as You
> received them, (b) in the description accompanying Your Submission, 
> include the phrase "Submission
> containing materials of a third party:" followed by the names of the 
> third party and any licenses or other
> restrictions of which You are aware, and (c) follow any other 
> instructions in the Project's written
> guidelines concerning Submissions.
> * Your Employer. References to "employer" in this Agreement include 
> Your employer or anyone else
> for whom You are acting in making Your Submission, e.g. as a 
> contractor, vendor, or agent. If Your
> Submission is made in the course of Your work for an employer or Your 
> employer has intellectual
> property rights in Your Submission by contract or applicable law, You 
> must secure permission from Your
> employer to make the Submission before signing this Agreement. In that 
> case, the term "You" in this
> Agreement will refer to You and the employer collectively. If You 
> change employers in the future and
> desire to Submit additional Submissions for the new employer, then You 
> agree to sign a new Agreement
> and secure permission from the new employer before Submitting those 
> Submissions.     * Licenses.
> 
> * Copyright License. You grant Microsoft, and those who receive the 
> Submission directly or
> indirectly from Microsoft, a perpetual, worldwide, non-exclusive, 
> royalty-free, irrevocable license in the
> Submission to reproduce, prepare derivative works of, publicly display, 
> publicly perform, and distribute
> the Submission and such derivative works, and to sublicense any or all 
> of the foregoing rights to third
> parties.
> * Patent License. You grant Microsoft, and those who receive the 
> Submission directly or
> indirectly from Microsoft, a perpetual, worldwide, non-exclusive, 
> royalty-free, irrevocable license under
> Your patent claims that are necessarily infringed by the Submission or 
> the combination of the
> Submission with the Project to which it was Submitted to make, have 
> made, use, offer to sell, sell and
> import or otherwise dispose of the Submission alone or with the 
> Project.
> * Other Rights Reserved. Each party reserves all rights not expressly 
> granted in this Agreement.
> No additional licenses or rights whatsoever (including, without 
> limitation, any implied licenses) are
> granted by implication, exhaustion, estoppel or otherwise.
> 
> * Representations and Warranties. You represent that You are legally 
> entitled to grant the above
> licenses. You represent that each of Your Submissions is entirely Your 
> original work (except as You may
> have disclosed under Section 3). You represent that You have secured 
> permission from Your employer to
> make the Submission in cases where Your Submission is made in the 
> course of Your work for Your
> employer or Your employer has intellectual property rights in Your 
> Submission by contract or applicable
> law. If You are signing this Agreement on behalf of Your employer, You 
> represent and warrant that You
> have the necessary authority to bind the listed employer to the 
> obligations contained in this Agreement.
> You are not expected to provide support for Your Submission, unless You 
> choose to do so. UNLESS
> REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING, AND EXCEPT FOR THE 
> WARRANTIES
> EXPRESSLY STATED IN SECTIONS 3, 4, AND 6, THE SUBMISSION PROVIDED UNDER 
> THIS AGREEMENT IS
> PROVIDED WITHOUT WARRANTY OF ANY KIND, INCLUDING, BUT NOT LIMITED TO, 
> ANY WARRANTY OF
> NONINFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
> * Notice to Microsoft. You agree to notify Microsoft in writing of any 
> facts or circumstances of which
> You later become aware that would make Your representations in this 
> Agreement inaccurate in any
> respect.
> * Information about Submissions. You agree that contributions to 
> Projects and information about
> contributions may be maintained indefinitely and disclosed publicly, 
> including Your name and other
> information that You submit with Your Submission.
> * Governing Law/Jurisdiction. This Agreement is governed by the laws of 
> the State of Washington, and
> the parties consent to exclusive jurisdiction and venue in the federal 
> courts sitting in King County,
> Washington, unless no federal subject matter jurisdiction exists, in 
> which case the parties consent to
> exclusive jurisdiction and venue in the Superior Court of King County, 
> Washington. The parties waive all
> defenses of lack of personal jurisdiction and forum non-conveniens.
> * Entire Agreement/Assignment. This Agreement is the entire agreement 
> between the parties, and
> supersedes any and all prior agreements, understandings or 
> communications, written or oral, between
> the parties relating to the subject matter hereof. This Agreement may 
> be assigned by Microsoft.
> 
> --
> Reply to this email directly, view it on GitHub [3], or unsubscribe 
> [4].
> You are receiving this because you were mentioned.Message ID: 
> ***@***.***>



Links:
------
[1] https://github.com/LetsDoSQL-jdw
[2] http://www.opensource.org
[3] 
https://github.com/microsoft/tigertoolbox/pull/314#issuecomment-2343177143
[4] 
https://github.com/notifications/unsubscribe-auth/AQPTAQLWCUY2BEQZNAEJBWDZWAHKJAVCNFSM6AAAAABOASQJGSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDGNBTGE3TOMJUGM

bluefooted commented on AdaptiveIndexDefrag/usp_AdaptiveIndexDefrag.sql:
```diff
@@ -1,8 +1,8 @@
 
--- For deployment in Azure SQL Database, remove or comment the USE statement below.
-USE msdb
+-- For deployment in Azure SQL Database, remove or comment the USE statement below. And please, please don't create user objects in system databases like [msdb].
```
Myeah.... I sort of agree with you when it comes to master, but msdb is a little more flexible. It's designed to store maintenance functions and metadata, and it actually needs to be user maintained or it can get out of control (something a lot of folks don't pay attention to). Also, if we were to change this here, we would have to add code to create this database elsewhere and then change all of the other scripts to use it. I'd prefer to keep this in msdb. All the other changes look good though, so if you can pull this one, I will merge the request.

LetsDoSQL-jdw replied:
Thanks for taking a look @bluefooted :) When I get back from holiday next week I will implement your proposal.

LetsDoSQL-jdw replied:
@bluefooted Leave AdaptiveIndexDefrag deployed to [msdb] database. Also, I noticed that the DateDiff_Big makes that the script will be SQL2016+

LetsDoSQL-jdw committed:
Update usp_AdaptiveIndexDefrag.sql

As @bluefooted requested, leave AdaptiveIndexDefrag stored in msdb for flexibility purposes

LetsDoSQL-jdw committed:
Update README.md

Update minimum version to SQL2016 because of datediff_big

thesqlsith commented on AdaptiveIndexDefrag/usp_AdaptiveIndexDefrag.sql:
```diff
@@ -1463,26 +1465,26 @@ CHAR(10) + 'WHERE mst.is_ms_shipped = 0 ' + CASE WHEN @dbScope IS NULL AND @tblN
 						BEGIN
 							INSERT INTO dbo.tbl_AdaptiveIndexDefrag_Working (dbID, dbName, objectID, indexID, partitionNumber, fragmentation, page_count, range_scan_count, record_count, scanDate)		
-							SELECT @dbID AS [dbID], QUOTENAME(DB_NAME(ps.database_id)) AS [dbName], @objectID AS [objectID], @indexID AS [indexID], ps.partition_number AS [partitionNumber], SUM(ps.avg_fragmentation_in_percent) AS [fragmentation], SUM(ps.page_count) AS [page_count], os.range_scan_count, ps.record_count, GETDATE() AS [scanDate]
+							SELECT @dbID AS [dbID], DB_NAME(ps.database_id) AS [dbName], @objectID AS [objectID], @indexID AS [indexID], ps.partition_number AS [partitionNumber], SUM(ps.avg_fragmentation_in_percent) AS [fragmentation], SUM(ps.page_count) AS [page_count], os.range_scan_count, ps.record_count, GETDATE() AS [scanDate]
```
It may still be a good idea to keep the QUOTENAME function in the context of user database names.

LetsDoSQL-jdw replied:
Here the query statement adds quotations to the values, which in line 1480 are to be stored into table. 

The bug is that most queries add quotations to the resultset when the queried data already have quotations. 

To fix that, I consistently _store_ object names to tables without quotations. Now all data in the tables are stored without quotations, except if there are old records, for example in the Exceptions-table.

For dynamic queries running _outside_ current database scope the quotenames are to be added to database names to make the statements correct, especially when those names have spaces. But again, not to store data to the tables. 

Also, quotenames are to be added for debugging and logging purposes to avoid confusion between keywords and object names.

thesqlsith commented on AdaptiveIndexDefrag/usp_AdaptiveIndexDefrag.sql:
```diff
@@ -1463,26 +1465,26 @@ CHAR(10) + 'WHERE mst.is_ms_shipped = 0 ' + CASE WHEN @dbScope IS NULL AND @tblN
 								AND ps.page_count >= @minPageCount
 								AND ps.index_level = 0 -- leaf-level nodes only, supports @scanMode
-							GROUP BY ps.database_id, QUOTENAME(DB_NAME(ps.database_id)), ps.partition_number, os.range_scan_count, ps.record_count	
+							GROUP BY ps.database_id, DB_NAME(ps.database_id), ps.partition_number, os.range_scan_count, ps.record_count	
```
It may still be a good idea to keep the QUOTENAME function in the context of user database names.

LetsDoSQL-jdw replied:
line 1487 is part of the same query started in line 1481, so the same change in GROUP BY is needed to make the entire statement work.

LetsDoSQL-jdw commented on AdaptiveIndexDefrag/usp_AdaptiveIndexDefrag.sql:
```diff
@@ -1463,26 +1465,26 @@ CHAR(10) + 'WHERE mst.is_ms_shipped = 0 ' + CASE WHEN @dbScope IS NULL AND @tblN
 						BEGIN
 							INSERT INTO dbo.tbl_AdaptiveIndexDefrag_Working (dbID, dbName, objectID, indexID, partitionNumber, fragmentation, page_count, range_scan_count, record_count, scanDate)		
-							SELECT @dbID AS [dbID], QUOTENAME(DB_NAME(ps.database_id)) AS [dbName], @objectID AS [objectID], @indexID AS [indexID], ps.partition_number AS [partitionNumber], SUM(ps.avg_fragmentation_in_percent) AS [fragmentation], SUM(ps.page_count) AS [page_count], os.range_scan_count, ps.record_count, GETDATE() AS [scanDate]
+							SELECT @dbID AS [dbID], DB_NAME(ps.database_id) AS [dbName], @objectID AS [objectID], @indexID AS [indexID], ps.partition_number AS [partitionNumber], SUM(ps.avg_fragmentation_in_percent) AS [fragmentation], SUM(ps.page_count) AS [page_count], os.range_scan_count, ps.record_count, GETDATE() AS [scanDate]
```
Hi @thesqlsith, thanks for taking the time to review my PR. After I created this PR, I made a few additional updates, which I brought live in production, but waited to commit them here until this PR is merged.

-------
Here the query statement adds quotations to the values, which in line 1467 are to be stored into table. 

The bug is that most queries add quotations to the resultset when the queried data already have quotations. 

To fix that, I consistently _store_ object names to tables without quotations. Now all data in the tables are stored without quotations, except if there are old records, for example in the Exceptions-table.

For dynamic queries running _outside_ current database scope the quotenames are to be added to database names to make the statements correct, especially when those names have spaces. But again, not to store data to the tables. 

Also, quotenames are to be added for debugging and logging purposes to avoid confusion between keywords and object names.

LetsDoSQL-jdw commented on AdaptiveIndexDefrag/usp_AdaptiveIndexDefrag.sql:
```diff
@@ -1463,26 +1465,26 @@ CHAR(10) + 'WHERE mst.is_ms_shipped = 0 ' + CASE WHEN @dbScope IS NULL AND @tblN
 								AND ps.index_level = 0 -- leaf-level nodes only, supports @scanMode
 								AND ps.alloc_unit_type_desc = 'IN_ROW_DATA'  -- exclude blobs
-							GROUP BY ps.database_id, QUOTENAME(DB_NAME(ps.database_id)), ps.partition_number, os.range_scan_count, ps.record_count	
+							GROUP BY ps.database_id, DB_NAME(ps.database_id), ps.partition_number, os.range_scan_count, ps.record_count	
```
line 1475 is part of the same query started in line 1468, so the same change in GROUP BY is needed to make the entire statement work.

LetsDoSQL-jdw commented on AdaptiveIndexDefrag/usp_AdaptiveIndexDefrag.sql:
```diff
@@ -1536,7 +1538,7 @@ CHAR(10) + 'WHERE mst.is_ms_shipped = 0 ' + CASE WHEN @dbScope IS NULL AND @tblN
 						BEGIN TRY
 							SELECT @ColumnStoreGetIXSQL = CASE WHEN @engineedition NOT IN (5, 6) THEN 'USE [' + DB_NAME(@dbID) + '];
-' ELSE '' END + 'SELECT @dbID_In, QUOTENAME(DB_NAME(@dbID_In)), rg.object_id, rg.index_id, rg.partition_number, CASE WHEN SUM((ISNULL(rg.deleted_rows,1)*100)/rg.total_rows) > 100 THEN 100 ELSE SUM((ISNULL(rg.deleted_rows,1)*100)/rg.total_rows) END AS [fragmentation], SUM(ISNULL(rg.size_in_bytes,1)/1024/8) AS [simulated_page_count], SUM(rg.total_rows) AS total_rows, GETDATE() AS [scanDate]	
+' ELSE '' END + 'SELECT @dbID_In, DB_NAME(@dbID_In), rg.object_id, rg.index_id, rg.partition_number, CASE WHEN SUM((ISNULL(rg.deleted_rows,1)*100)/rg.total_rows) > 100 THEN 100 ELSE SUM((ISNULL(rg.deleted_rows,1)*100)/rg.total_rows) END AS [fragmentation], SUM(ISNULL(rg.size_in_bytes,1)/1024/8) AS [simulated_page_count], SUM(rg.total_rows) AS total_rows, GETDATE() AS [scanDate]	
```
Here the dynamic query adds quotations to the values, which in line 1552 & 1553 are to be stored into table. 

The bug is that most queries add quotations to the resultset when the queried data already have quotations. 

To fix that, I consistently _store_ object names to tables without quotations. Now all data in the tables are stored without quotations, except if there are old records, for example in the Exceptions-table.

For dynamic queries running _outside_ current database scope the quotenames are to be added to database names to make the statements correct, especially when those names have spaces. But again, not to store data to the tables. 

Also, quotenames are to be added for debugging and logging purposes to avoid confusion between keywords and object names.

LetsDoSQL-jdw commented on AdaptiveIndexDefrag/usp_AdaptiveIndexDefrag.sql:
```diff
@@ -1582,7 +1584,7 @@ OPTION (MAXDOP 2)'
 				/* Look up index status for various purposes */	
 				SELECT @updateSQL = N'UPDATE ids		
-SET schemaName = QUOTENAME(s.name), objectName = QUOTENAME(o.name), indexName = QUOTENAME(i.name), is_primary_key = i.is_primary_key, fill_factor = i.fill_factor, is_disabled = i.is_disabled, is_padded = i.is_padded, is_hypothetical = i.is_hypothetical, has_filter = ' + CASE WHEN @sqlmajorver >= 10 THEN 'i.has_filter' ELSE '0' END + ', allow_page_locks = i.allow_page_locks, type = i.type
+SET schemaName = s.name, objectName = o.name, indexName = i.name, is_primary_key = i.is_primary_key, fill_factor = i.fill_factor, is_disabled = i.is_disabled, is_padded = i.is_padded, is_hypothetical = i.is_hypothetical, has_filter = ' + CASE WHEN @sqlmajorver >= 10 THEN 'i.has_filter' ELSE '0' END + ', allow_page_locks = i.allow_page_locks, type = i.type
```
Here the dynamic query adds quotations to the values to be stored into table.

The bug is that most queries add quotations to the resultset when the queried data already have quotations. 

To fix that, I consistently _store_ object names to tables without quotations. Now all data in the tables are stored without quotations, except if there are old records, for example in the Exceptions-table.

For dynamic queries running _outside_ current database scope the quotenames are to be added to database names to make the statements correct, especially when those names have spaces. But again, not to store data to the tables. 

Also, quotenames are to be added for debugging and logging purposes to avoid confusion between keywords and object names.

LetsDoSQL-jdw commented on AdaptiveIndexDefrag/usp_AdaptiveIndexDefrag.sql:
```diff
@@ -1635,8 +1637,8 @@ WHERE ids.[dbID] = ' + CAST(@dbID AS NVARCHAR(10));
-' ELSE '' END + 'SELECT DISTINCT ' + CAST(@dbID AS NVARCHAR(10)) + ', ''[' + DB_NAME(@dbID) + ']'', ss.[object_id], ss.stats_id, ' + CASE WHEN ((@sqlmajorver = 12 AND @sqlbuild >= 5000) OR @sqlmajorver > 12) THEN 'ISNULL(sp.partition_number,1),' ELSE '1,' END + '
-	QUOTENAME(s.name), QUOTENAME(so.name), QUOTENAME(ss.name), ss.[no_recompute], ' + CASE WHEN @sqlmajorver < 12 THEN '0 AS ' ELSE 'ss.' END + '[is_incremental], GETDATE() AS scanDate
+' ELSE '' END + 'SELECT DISTINCT ' + CAST(@dbID AS NVARCHAR(10)) + ', ' + QUOTENAME(DB_NAME(@dbID),'''') + ' AS dbName, ss.[object_id], ss.stats_id, ' + CASE WHEN ((@sqlmajorver = 12 AND @sqlbuild >= 5000) OR @sqlmajorver > 12) THEN 'ISNULL(sp.partition_number,1),' ELSE '1,' END + '
+	s.name, so.name, ss.name, ss.[no_recompute], ' + CASE WHEN @sqlmajorver < 12 THEN '0 AS ' ELSE 'ss.' END + '[is_incremental], GETDATE() AS scanDate
```
Here the dynamic query adds quotations to the values to be stored into table.

The bug is that most queries add quotations to the resultset when the queried data already have quotations. 

To fix that, I consistently _store_ object names to tables without quotations. Now all data in the tables are stored without quotations, except if there are old records, for example in the Exceptions-table.

For dynamic queries running _outside_ current database scope the quotenames are to be added to database names to make the statements correct, especially when those names have spaces. But again, not to store data to the tables. 

Also, quotenames are to be added for debugging and logging purposes to avoid confusion between keywords and object names.
